2019/d07Amplifiers.d

```diff
@@ -0,0 +1,131 @@
+auto thrusterSignal(int[] program, int[] phases) {
+  import intcode;
+  int input;
+
+  foreach(phase; phases) {
+    input = IntCode(program.dup).run([phase, input])[0];
+  }
+
+  return input;
+}
+
+auto findMaxThrusterSignal(T)(int[] program, int phase, T callback) {
+  import std.algorithm, std.range, std.typecons;
+
+  int maxSignal;
+  int[] maxPhases;
+
+  foreach(phases; iota(phase, phase+5).permutations) {
+    int signal = callback(program, phases.array);
+    if (signal > maxSignal) {
+      maxSignal = signal;
+      maxPhases = phases.array;
+    }
+  }
+
+  return tuple(maxSignal, maxPhases);
+}
+
+auto maxThrusterSignal(int[] program) {
+  return findMaxThrusterSignal(program, 0, &thrusterSignal);
+}
+
+auto thrusterSignalWithFeedback(int[] program, int[] phases) {
+  import intcode;
+
+  IntCode[] amps;
+  foreach(phase; phases) {
+    IntCode ic = IntCode(program.dup);
+    ic.runInFeedback([phase]);
+    amps ~= ic;
+  }
+
+  auto feedback = [0];
+  bool halted;
+
+  do {
+    halted = true;
+
+    foreach(ref amp; amps) {
```

Comment:
For posterity, without `ref` here, the amps went into an endless feedback loop because their internal view of `halted` was not updated in the array original.